### ISSUE TITLE
Add "insert_zero" option for np.diff

### ISSUE THREAD

**Issue by petered** [numpy/numpy#8132]
I often find myself wanting the difference of an array along some axis, where the 0th element should just be the first entry of the array along that axis.  Currently, I do this with the following awkward construction:

```
a = np.random.RandomState(1234).randint(12, size=(3, 4))
array([[ 3,  6,  5,  4],
       [ 8,  9,  1,  7],
       [ 9, 10, 11, 10]])
# Want a diff along columns of a:
np.diff(np.insert(a, 0, 0, axis=0), axis=0)
> array([[ 3,  6,  5,  4],
       [ 5,  3, -4,  3],
       [ 1,  1, 10,  3]])
# It would be more convenient to go:
np.diff(a, axis=0, insert_zero=True)
```

This would allow us to have an easy inverse of cumsum, so that

`np.all( np.cumsum(np.diff(arr, axis=1, insert_zero=True)) == arr)`

**Comment by shoyer** [comment]
The inverse of cumsum problem would also be solved by adding `include_identity` to cumsum (#6044), which is something I've felt a stronger need for.

**Comment by mattharrigan** [comment]
I prefer the ediff1d arguments to_begin and to_end, allowing for non zero and/or multiple elements to be inserted on either side.

**Comment by petered** [comment]
Maybe the best solution is to add the padding options of ediff1d to diff, and then kill ediff1d.  The discussion [here](http://stackoverflow.com/questions/39014324/difference-between-numpy-ediff1d-and-diff) does kind of make it seem like ediff1d should go.

**Comment by mattharrigan** [comment]
I agree, that's basically the conclusion of part of the discussion of this [issue](https://github.com/numpy/numpy/issues/8175) also

**Comment by bashtage** [comment]
A general initial condition would be better than a specific one.

**Comment by petered** [comment]
Agreed.

**Comment by petered** [comment]
Horray!  Thanks guys.
